java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java

```diff
@@ -0,0 +1,920 @@
+        case CSRC:
+          OrtUtil.BufferTuple innerIndicesTuple =
+              OrtUtil.prepareBuffer(((CSRCTensor) tensor).getInnerIndices(), indicesType);
+          return new OnnxSparseTensor(
+              createCSRCSparseTensorFromBuffer(
+                  OnnxRuntime.ortApiHandle,
+                  allocator.handle,
+                  indicesTuple.data,
+                  indicesTuple.pos,
+                  indicesTuple.size,
+                  innerIndicesTuple.data,
+                  innerIndicesTuple.pos,
+                  innerIndicesTuple.size,
+                  valuesTuple.data,
+                  valuesTuple.pos,
+                  info.shape,
+                  tensor.getValuesShape(),
+                  info.onnxType.value),
+              allocator.handle,
+              tensor.getSparsityType(),
+              info,
+              indicesTuple.data,
+              (LongBuffer) innerIndicesTuple.data,
+              valuesTuple.data);
+        case UNDEFINED:
+        default:
+          throw new IllegalArgumentException("Cannot create an UNDEFINED sparse tensor.");
+      }
+    } else {
+      throw new IllegalStateException(
+          "Trying to create an OnnxSparseTensor on a closed OrtAllocator.");
+    }
+  }
+
+  @Override
+  public OnnxValueType getType() {
+    return OnnxValueType.ONNX_TYPE_SPARSETENSOR;
+  }
+
+  @Override
+  public SparseTensor<? extends Buffer> getValue() throws OrtException {
+    Buffer buffer = getValuesBuffer();
+    long[] indicesShape = getIndicesShape(OnnxRuntime.ortApiHandle, nativeHandle);
+    switch (sparseTensorType) {
+      case COO:
+        return new COOTensor(
+            (LongBuffer) getIndicesBuffer(),
+            indicesShape,
+            buffer,
+            info.shape,
+            info.type,
+            buffer.remaining());
+      case CSRC:
+        return new CSRCTensor(
+            (LongBuffer) getIndicesBuffer(),
+            getInnerIndicesBuffer(),
+            buffer,
+            info.shape,
+            info.type,
+            buffer.remaining());
+      case BLOCK_SPARSE:
+        long[] valuesShape = getValuesShape(OnnxRuntime.ortApiHandle, nativeHandle);
+        return new BlockSparseTensor(
+            (IntBuffer) getIndicesBuffer(),
+            indicesShape,
+            buffer,
+            valuesShape,
+            info.shape,
+            info.type,
+            buffer.remaining());
+      case UNDEFINED:
+      default:
+        throw new IllegalStateException("Undefined sparsity type in this sparse tensor.");
+    }
+  }
+
+  @Override
+  public void close() {
+    close(OnnxRuntime.ortApiHandle, nativeHandle);
+  }
+
+  /**
+   * Returns the type of this OnnxSparseTensor.
+   *
+   * @return The sparsity type.
+   */
+  public SparseTensorType getSparseTensorType() {
+    return sparseTensorType;
+  }
+
+  /**
+   * Gets a copy of the indices.
+   *
+   * <p>These are the outer indices if it's a CSRC sparse tensor.
```

Comment:
>p>These are the outer indices if it's a CSRC sparse tensor. [](http://example.com/codeflow?start=6&length=59)

I would expect to get all the indices to be able to do anything meaningful with them. Why inner indices are not retrieved? I think this interface is not well thought out.

Reply:
They are accessed in two different ways for the CSRC tensor, as there are two methods that return them in the C API. This mirrors that into Java. We can't easily have the Java code return different numbers of things dependent on the native state, so in the CSRC case users need to do two calls to get the two buffers.